src/pages/experiments/[experimentId]/data-processing/configure-embedding/components/EmbeddingPreview.jsx

```diff
@@ -40,99 +41,131 @@ import LabelsDesign from '../../../plots-and-tables/components/LabelsDesign';
+      return;
+    }
 
-    if (!config) {
+    if (!selectedExpression && config.shownGene) {
+      dispatch(loadGeneExpression(experimentId, [config.shownGene]));
+    }
+
+    if ((plots[selectedSpec].plotType === 'embeddingContinuous' && !selectedExpression)
+      || !cellSets) {
       return;
     }
 
     const spec = plots[selectedSpec].specGenerator(config);
-    generateData(spec);
+
+    if (plots[selectedSpec].plotType === 'embeddingContinuous') {
+      // If embedding is continuous
+
+      spec.data.forEach((s) => {
+        if (s.name === 'expression') {
+          s.values = selectedExpression;
+        } else if (s.name === 'embedding') {
+          s.values = data;
+        }
+      });
+    } else {
+      // If embedding categorical
+      // First, find the child nodes in the hirerarchy.
+      let newCellSets = cellSets.hierarchy.find(
+        (rootNode) => rootNode.key === config.selectedCellSet,
+      )?.children || [];
+
+      // Build up the data source based on the properties. Note that the child nodes
+      // in the hierarchy are /objects/ with a `key` property, hence the destructuring
+      // in the function.
+      newCellSets = newCellSets.map(({ key }) => ({
+        cellSetId: key,
+        ...cellSets.properties[key],
+        cellIds: Array.from(cellSets.properties[key].cellIds),
+      }));
+
+      spec.data.forEach((s) => {
+        if (s.name === 'cellSets') {
+          s.values = newCellSets;
+        } else if (s.name === 'embedding') {
+          s.values = data;
+        }
+      });
+    }
+
     setPlotSpec(spec);
-  }, [config]);
+  }, [config, data, selectedExpression, cellSets]);
 
   // If the user toggles to a different embedding, set the config to be the initial
   // state for that type of plot.
   useEffect(() => {
-    setConfig(plots[selectedSpec].initialConfig);
-  }, [selectedSpec]);
+    if (!isBrowser) return;
+    const { plotUuid, plotType } = plots[selectedSpec];
 
-  // Quick and dirty function to massage prepared data into a good shape.
-  // This will be changed once we actually load data from Redux.
-  /* eslint-disable no-param-reassign */
-  const generateData = (spec) => {
-    spec.data.forEach((s) => {
-      if (s.name === 'cellSets') {
-        s.values = [];
-
-        UMAP.forEach((cell, i) => {
-          s.values[cell.cluster_id] = {
-            name: `${cell.cluster_id}`,
-            cellSetId: cell.cluster_id,
-            cellIds: s.values[cell.cluster_id]?.cellIds ? [...s.values[cell.cluster_id].cellIds, i] : [i],
-            color: colorProvider.getColor(),
-          };
-        });
-      }
-
-      if (s.name === 'expression') {
-        s.values = { expression: UMAP.map((cell) => cell.doubletScore || 0) };
-      }
-
-      if (s.name === 'embedding') {
-        s.values = UMAP.map((cell) => [cell.UMAP_1, cell.UMAP_2]);
-      }
-    });
-  };
+    if (!config) {
+      dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
```

Comment:
we just found a bug - we should change `loadPlotConfig` to pass in `plotUuid` on line 26 instead of plotType

Reply:
I just went through the code again, and find out it's not a bug. We misunderstood it the first time. The code is used to fill the config with `initialConfig` of a given `plotType` if `plotUuid` is not found (hence the 404). I tried changing the `plotType` with `plotUuid` and it doesn't work, so I returned it to `plotType`.